modules/imgproc/src/canny.cpp

```diff
@@ -605,20 +618,103 @@ class tbbCanny
             if (!m[mapstep])    CANNY_PUSH(m + mapstep);
             if (!m[mapstep+1])  CANNY_PUSH(m + mapstep + 1);
         }
+
+        AutoLock lock(*mutex);
+        while (!borderPeaksLocal.empty()) {
+            _borderPeaksParallel->push(borderPeaksLocal.front());
+            borderPeaksLocal.pop();
+        }
     }
 
 private:
-    const Range boundaries;
     const Mat& src;
     uchar* map;
-    int low;
-    int high;
-    int aperture_size;
+    int low, high, aperture_size;
     bool L2gradient;
+    std::queue<uchar*> *_borderPeaksParallel;
+    Mutex *mutex;
```

Comment:
It is because of "const" modifier:
```
void operator()(const Range &boundaries) const
```

This should work fine:
```
mutable Mutex mutex;
```